ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,862 @@
+    }
+    Ok(hints_bag)
+}
+
+/// Extracting hints from a transaction and outputs it's corresponding TransactionHintsBag
+pub fn extract_hints(
+    signed_tx: Transaction,
+    tx_context: TransactionContext,
+    state_context: &ErgoStateContext,
+    real_secrets_to_extract: Vec<SigmaBoolean>,
+    simulated_secrets_to_extract: Vec<SigmaBoolean>,
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let mut hints_bag = TransactionHintsBag::empty();
+    let tx = tx_context.spending_tx.clone();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = tx_context
+            .get_boxes_to_spend()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let ctx = Rc::new(make_context(state_context, &tx_context, i)?);
+        let tree = input_box.ergo_tree.clone();
+        let proof: ProofBytes = signed_tx
+            .inputs
+            .get(i)
+            .unwrap()
+            .clone()
+            .spending_proof
+            .proof;
+        let proof: Vec<u8> = Vec::from(proof);
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(
+            i,
+            bag_for_multi_sig(
+                sigma_tree,
+                real_secrets_to_extract.as_slice(),
+                simulated_secrets_to_extract.as_slice(),
+                proof.as_slice(),
+            ),
+        );
+    }
+
+    Ok(hints_bag)
+}
+
+#[allow(clippy::unwrap_used)]
+/// Extracting hints from reduced transaction
+pub fn extract_hints_from_reduced_transaction(
+    reduced_tx: ReducedTransaction,
+    signed_tx: Transaction,
+    real_propositions: Vec<SigmaBoolean>,
+    simulated_propositions: Vec<SigmaBoolean>,
+) -> TransactionHintsBag {
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (idx, input) in reduced_tx.reduced_inputs().iter().enumerate() {
+        let sigma_prop = input.clone().reduction_result.sigma_prop;
+        let proof = Vec::from(
+            signed_tx
+                .inputs
+                .get(idx)
+                .unwrap()
+                .spending_proof
+                .clone()
+                .proof,
+        );
+        let hints = bag_for_multi_sig(
+            sigma_prop,
+            real_propositions.as_slice(),
+            simulated_propositions.as_slice(),
+            proof.as_slice(),
+        );
+        hints_bag.add_hints_for_input(idx, hints);
+    }
+    hints_bag
+}
+
+// The unwrap use on option causes panic in case of dhtuple signature, that is not implemented
+#[allow(clippy::unwrap_used)]
```

Comment:
Please, handle the errors without unwrapping here. In general, unwraps besides tests are allowed for exceptional cases only.